### PR TITLE
[xds] xDS client per data plane target

### PR DESCRIPTION
**Pull request by eugeneo**

### CODE REVIEW

**Review by markdroth — COMMENTED** [review]
The overall structure here looks good!  Comments are mostly minor, but there are some edge cases that we need to fix.

Please let me know if you have any questions.  Thanks!

File: test/cpp/end2end/xds/xds_core_end2end_test.cc [hunk 1]
```diff
@@ -182,7 +182,7 @@ TEST_P(GlobalXdsClientTest, MultipleChannelsShareXdsClient) {
   channel2->GetState(/*try_to_connect=*/true);
   ASSERT_TRUE(channel2->WaitForConnected(grpc_timeout_seconds_to_deadline(1)));
   // Make sure there's only one client connected.
-  EXPECT_EQ(1UL, balancer_->ads_service()->clients().size());
+  EXPECT_EQ(2UL, balancer_->ads_service()->clients().size());
```
[the remark this answers is not in the document]
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_core_end2end_test.cc:199]: Discussed offline, split to two test cases with same and different targets.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 2]
```diff
@@ -344,14 +344,15 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpVanilla) {
...
-  EXPECT_EQ(1, csds_response.config_size());
+  ASSERT_EQ(1, csds_response.config_size());
   const auto& client_config = csds_response.config(0);
   // Validate the Node information
   EXPECT_THAT(client_config.node(),
               EqNode("xds_end2end_test", ::testing::HasSubstr("C-core"),
                      ::testing::HasSubstr(grpc_version_string()),
                      ::testing::ElementsAre(
                          "envoy.lb.does_not_support_overprovisioning")));
+  EXPECT_EQ(client_config.client_scope(), "server.example.com");
```
[the remark this answers is not in the document]
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_csds_end2end_test.cc]: Done

File: test/cpp/end2end/xds/xds_core_end2end_test.cc [hunk 3]
```diff
@@ -260,8 +260,6 @@ TEST_P(GlobalXdsClientTest, MultipleBadLdsResources) {
         response_state->error_message,
         "xDS response validation errors: ["
         "resource index 0: server.other.com: "
-        "INVALID_ARGUMENT: Listener has neither address nor ApiListener; "
```
[the remark this answers is not in the document]
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_core_end2end_test.cc:263]: Removed.

**Review by eugeneo — COMMENTED** [review]
Thank you for the review. I addressed the comments.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 4]
```diff
@@ -288,9 +278,7 @@ void SetXdsChannelArgsForTest(grpc_channel_args* args) {
 }
 
 void UnsetGlobalXdsClientsForTest() {
-  MutexLock lock(g_mu);
-  delete g_xds_client_map;
-  g_xds_client_map = nullptr;
+  // TODO(eostroukhov): Remove before submit
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:281]: I think we should clear the map here.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 5]
```diff
@@ -94,8 +94,8 @@ namespace {
 Mutex* g_mu = new Mutex;
 const grpc_channel_args* g_channel_args ABSL_GUARDED_BY(*g_mu) = nullptr;
 // Key bytes live in clients so they outlive the entries in this map
-std::map<absl::string_view, WeakRefCountedPtr<GrpcXdsClient>, std::less<>>*
-    g_xds_client_map ABSL_GUARDED_BY(*g_mu) = nullptr;
+NoDestruct<std::map<absl::string_view, GrpcXdsClient*, std::less<>>>
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:97]: I don't think we need the `std::less<>` here.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 6]
```diff
@@ -267,11 +259,9 @@ GrpcXdsClient::GrpcXdsClient(
 
 void GrpcXdsClient::Orphan() {
   MutexLock lock(g_mu);
-  if (g_xds_client_map != nullptr && g_xds_client_map->size() > 1) {
+  auto it = g_xds_client_map->find(key_);
+  if (it != g_xds_client_map->end() && it->second == this) {
     g_xds_client_map->erase(g_xds_client_map->find(key_));
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:264]: This should be `g_xds_client_map->erase(it)`.  We've already done the lookup, so there's no reason to do it again.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 7]
```diff
@@ -284,6 +285,51 @@ class ClientStatusDiscoveryServiceTest : public XdsEnd2endTest {
     return response;
   }
 
+  std::vector<::testing::Matcher<
+      envoy::service::status::v3::ClientConfig_GenericXdsConfig>>
+  PrepareClientConfigMatchers(bool enable_rds_testing,
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:290]: Please don't refactor test data out of the individual tests, since it makes it harder to understand what is happening in the test without reading code outside of the test.  If we have to duplicate this block of code in two tests, that's preferable.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 8]
```diff
@@ -616,6 +627,49 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpClusterRequested) {
                       ClientResourceStatus::REQUESTED, ::testing::_))));
 }
 
+TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpMultiClient) {
+  Listener listener = default_listener_;
+  listener.set_name("server2.example.com");
+  balancer_->ads_service()->SetLdsResource(listener);
+  SetListenerAndRouteConfiguration(balancer_.get(), listener,
+                                   default_route_config_);
+
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:636]: Nit: Please remove blank line.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 9]
```diff
@@ -616,6 +627,49 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpClusterRequested) {
                       ClientResourceStatus::REQUESTED, ::testing::_))));
 }
 
+TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpMultiClient) {
+  Listener listener = default_listener_;
+  listener.set_name("server2.example.com");
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:632]: Please create a local string constant for this string, since we use it in two places.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h [hunk 10]
```diff
@@ -29,6 +29,7 @@
 #include "absl/strings/str_cat.h"
 #include "absl/strings/string_view.h"
 #include "absl/types/optional.h"
+#include "google/protobuf/message.h"
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_end2end_test_lib.h:32]: Why are we adding this include?  I don't see any changes in this file that require proto.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h [hunk 11]
```diff
@@ -164,6 +165,10 @@ class XdsTestType {
     return info.param.AsString();
   }
 
+  friend void PrintTo(const XdsTestType& test_type, std::ostream* os) {
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_end2end_test_lib.h:168]: Why is this needed?  We should be using the `Name()` method whenever we call `INSTANTIATE_TEST_SUITE_P()`.

File: test/cpp/end2end/xds/xds_core_end2end_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply test/cpp/end2end/xds/xds_core_end2end_test.cc:263]: I just realized that there is still a case where there can be multiple LDS resources in a single XdsClient, which is if there is a server listening on multiple ports (or just multiple servers, each listening on a separate port).  So let's add a test for that case to ensure that we report all errors in the NACK.

**Review by markdroth — COMMENTED** [review]
This looks really good!  Comments are mostly minor.

Please let me know if you have any questions.  Thanks!

File: test/cpp/end2end/xds/xds_core_end2end_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_core_end2end_test.cc:263]: I updated the test but there's race conditions where sometimes there's only a single LDS in the response, sometimes there LDS resources are in different order. I would really appreciate a suggestion how to test this.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc (hunk 7 above)
Comment (markdroth) [reply test/cpp/end2end/xds/xds_csds_end2end_test.cc:290]: I don't see why we need this helper in the new test.  I think we can construct the list of matchers inside the new test, just like we're doing with the existing test.

**Review by markdroth — COMMENTED** [review]
The production code looks great!  Remaining comments are only in the tests.

Please let me know if you have any questions.  Thanks!

File: test/cpp/end2end/xds/xds_end2end_test_lib.h (hunk 10 above)
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h]: Done. Unless IWYU decides to bring it back...

File: test/cpp/end2end/xds/xds_end2end_test_lib.h (hunk 11 above)
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h:167]: It's for failing test:
No change:
```
[  FAILED  ] XdsTest/ClientStatusDiscoveryServiceTest.XdsConfigDumpMultiClient/V3BootstrapFromEnvVar, where GetParam() = 20-byte object <00-00 00-00 00-00 00-00 02-00 00-00 02-00 00-00 00-00 00-00>
```
With this change:
```
[  FAILED  ] XdsTest/ClientStatusDiscoveryServiceTest.XdsConfigDumpMultiClient/V3BootstrapFromEnvVar, where GetParam() = V3BootstrapFromEnvVar
```
I now see that filter can be setup, I did not notice there was a parameter value after the slash.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 12]
```diff
@@ -615,6 +661,49 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpClusterRequested) {
...
+    absl::ConsumePrefix(&server, "xds:/");
+    // Validate the dumped xDS configs
+    EXPECT_THAT(
+        client_config.generic_xds_configs(),
+        ::testing::UnorderedElementsAreArray(
+            PrepareClientConfigMatchers(GetParam().enable_rds_testing(), server,
+                                        backends_[0]->port(), "3", "2")));
+  }
+  EXPECT_THAT(scopes,
+              ::testing::UnorderedElementsAre("xds:/server.example.com",
```
[the remark this answers is not in the document]
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_csds_end2end_test.cc]: Done

**Review by eugeneo — COMMENTED** [review]
Thank you for the review!

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 13]
```diff
@@ -692,16 +647,52 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpMultiClient) {
     scopes.emplace_back(client_config.client_scope());
     absl::string_view server = client_config.client_scope();
     absl::ConsumePrefix(&server, "xds:/");
+
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:650]: Nit: Please remove blank line.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 14]
```diff
@@ -692,16 +647,52 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpMultiClient) {
...
+        };
+    // If RDS is enabled, add matcher for RDS resource.
+    if (GetParam().enable_rds_testing()) {
+      matchers.push_back(EqGenericXdsConfig(
+          kRdsTypeUrl, kDefaultRouteConfigurationName, "2",
+          UnpackRouteConfiguration(EqRouteConfiguration(
+              kDefaultRouteConfigurationName, kDefaultClusterName)),
+          ClientResourceStatus::ACKED, ::testing::_));
+    }
+
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:689]: Nit: Please remove blank line.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 15]
```diff
@@ -237,13 +248,17 @@ GrpcXdsClient::GrpcXdsClient(
                    args.GetDurationFromIntMillis(
                            GRPC_ARG_XDS_RESOURCE_DOES_NOT_EXIST_TIMEOUT_MS)
                        .value_or(Duration::Seconds(15)))),
+      key_(key),
       certificate_provider_store_(MakeOrphanable<CertificateProviderStore>(
           static_cast<const GrpcXdsBootstrap&>(this->bootstrap())
               .certificate_providers())) {}
 
-GrpcXdsClient::~GrpcXdsClient() {
+void GrpcXdsClient::Orphan() {
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:256]: I just noticed a bug: This needs to call the parent class (`XdsClient::Orphan()`) in addition to the work it's doing here.

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc [hunk 16]
```diff
@@ -615,6 +616,85 @@ TEST_P(ClientStatusDiscoveryServiceTest, XdsConfigDumpClusterRequested) {
...
+  for (const auto& client_config : csds_response.config()) {
+    // Validate the Node information
+    EXPECT_THAT(client_config.node(),
+                EqNode("xds_end2end_test", ::testing::HasSubstr("C-core"),
+                       ::testing::HasSubstr(grpc_version_string()),
+                       ::testing::ElementsAre(
+                           "envoy.lb.does_not_support_overprovisioning")));
+    scopes.emplace_back(client_config.client_scope());
+    absl::string_view server = client_config.client_scope();
+    absl::ConsumePrefix(&server, "xds:/");
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_csds_end2end_test.cc:649]: I don't think this is needed anymore.

File: test/cpp/end2end/xds/xds_end2end_test_lib.h (hunk 11 above)
Comment (markdroth) [reply test/cpp/end2end/xds/xds_end2end_test_lib.h:167]: The fact that there's a parameter after the slash is exactly why I don't think we need this.  The log line that you quoted shows that this string is already logged.

But I'll leave it up to you.

File: test/cpp/end2end/xds/xds_core_end2end_test.cc [hunk 17]
```diff
@@ -232,6 +236,75 @@ TEST_P(
...
+          "resource index 0: "
+          "grpc/server?xds.resource.listening_address=127.0.0.1:%lu: "
+          "INVALID_ARGUMENT: Listener has neither address nor ApiListener]",
+          backends_[0]->port()));
+  listener = default_server_listener_;
+  listener.clear_address();
+  listener.set_name(GetServerListenerName(backends_[1]->port()));
+  balancer_->ads_service()->SetLdsResource(listener);
+  backends_[1]->Start();
+  // Sort as the message will output resources in sorted order
```
[the remark this answers is not in the document]
Comment (markdroth) [reply test/cpp/end2end/xds/xds_core_end2end_test.cc]: image

**Review by markdroth — COMMENTED** [review]
I noticed one small bug in the production code, otherwise this looks very good!

File: test/cpp/end2end/xds/xds_csds_end2end_test.cc (hunk 16 above)
Comment (eugeneo) [reply test/cpp/end2end/xds/xds_csds_end2end_test.cc]: Yep. Removed.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 18]
```diff
@@ -164,13 +167,14 @@ absl::StatusOr<RefCountedPtr<GrpcXdsClient>> GrpcXdsClient::GetOrCreate(
         GRPC_ARG_TEST_ONLY_DO_NOT_USE_IN_PROD_XDS_CLIENT_CHANNEL_ARGS);
     auto channel_args = ChannelArgs::FromC(xds_channel_args);
     return MakeRefCounted<GrpcXdsClient>(
-        std::move(*bootstrap), channel_args,
+        std::string(key), std::move(*bootstrap), channel_args,
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:170]: No need to convert to `std::string` here, now that we're passing the parameter to the ctor as `absl::string_view`.

File: src/core/ext/xds/xds_client_grpc.cc [hunk 19]
```diff
@@ -188,9 +192,13 @@ absl::StatusOr<RefCountedPtr<GrpcXdsClient>> GrpcXdsClient::GetOrCreate(
   // Instantiate XdsClient.
   auto channel_args = ChannelArgs::FromC(g_channel_args);
   auto xds_client = MakeRefCounted<GrpcXdsClient>(
-      std::move(*bootstrap), channel_args,
+      std::string(key), std::move(*bootstrap), channel_args,
```
Comment (markdroth) [inline src/core/ext/xds/xds_client_grpc.cc:195]: Same here: No need to convert to `std::string`.

**Review by markdroth — APPROVED** [review]
This looks great!  I noticed one more minor issue, but feel free to merge after addressing that.
